rs/registry/canister/src/mutations/node_management/do_replace_operator.rs

```diff
@@ -0,0 +1,628 @@
+        ));
+
+        // Increment old operator allowance.
+        let old_node_operator_key = make_node_operator_record_key(old_operator_id);
+        let updated_old_operator_record = NodeOperatorRecord {
+            node_allowance: old_operator_record.node_allowance + node_mutations,
+            ..old_operator_record.clone()
+        };
+        valid_mutations.push(update(
+            old_node_operator_key,
+            updated_old_operator_record.encode_to_vec(),
+        ));
+
+        self.maybe_apply_mutation_internal(valid_mutations);
+
+        println!(
+            "{}do_replace_operator: Finished executing payload: {:?}",
+            LOG_PREFIX, payload
+        );
+
+        Ok(())
+    }
+
+    /// Return the set of node records that belong to `old_operator_id`.
+    ///
+    /// If the node record is linked to the `new_operator_id`, it will
+    /// be filtered from the results, meaning that not all nodes from
+    /// `provided_node_ids` have to be returned.
+    fn maybe_fetch_nodes_to_update(
+        &self,
+        provided_node_ids: &Vec<NodeId>,
+        new_operator_id: &PrincipalId,
+        old_operator_id: &PrincipalId,
+    ) -> Result<BTreeMap<NodeId, NodeRecord>, String> {
+        let mut node_records = BTreeMap::new();
+
+        for node_id in provided_node_ids {
+            let node_record = self.get_node(*node_id).ok_or_else(|| {
+                format!(
+                    "{}do_replace_operator: Node not found: {}",
+                    LOG_PREFIX, node_id
+                )
+            })?;
+
+            if node_record.node_operator_id.eq(new_operator_id.as_slice()) {
+                println!(
+                    "{}do_replace_operator: Node {} already belongs to node operator {}",
+                    LOG_PREFIX, node_id, new_operator_id
+                );
+                continue;
+            }
+
+            if node_record.node_operator_id.ne(old_operator_id.as_slice()) {
+                return Err(format!(
+                    "{}do_replace_operator: Node {} does not belong to node operator {}",
+                    LOG_PREFIX, node_id, old_operator_id
+                ));
+            }
+
+            node_records.insert(*node_id, node_record);
+        }
+
+        Ok(node_records)
+    }
+
+    /// Fetches all the node operator records for a single
+    /// node provider.
+    fn maybe_fetch_operators_for_provider(
+        &self,
+        provider_id: &PrincipalId,
+    ) -> Result<Vec<NodeOperatorRecord>, String> {
+        let operators: Vec<_> = self
+            .get_node_operators_and_dcs_of_node_provider(*provider_id)
+            .map(|operators_and_dcs| operators_and_dcs.into_iter().map(|(_, o)| o).collect())
+            .map_err(|e| format!("{}do_replace_operator: {:?}", LOG_PREFIX, e))?;
+
+        if operators.is_empty() {
+            return Err(format!(
+                "{}do_replace_operator: Unknown node provider {}",
```

Comment:
Does empty really mean "unknown"? Can't a node provider simply have no node operators?

Reply:
In our data model we only have `node_operator`s. The only mention of `node_provider` is in the [Node operator record](https://github.com/dfinity/ic/blob/master/rs/protobuf/src/gen/registry/registry.node_operator.v1.rs#L30). I don't know why we did like that but it was like this for quite some time. In the governance canister we do actually have node providers, but in the registry we don't. The function I use `get_node_operators_and_dcs_of_node_provider` only operates on the registry. This means that if we don't find any node operator records it means that we don't know that node provider (from the registry canister).